Wikipedia/Code/WebViewController.m

```diff
@@ -147,6 +150,34 @@ - (void)viewWillAppear:(BOOL)animated {
                                                object:nil];
     // should happen in will appear to prevent bar from being incorrect during transitions
     [self updateZeroState];
+
+    @weakify(self);
+    void (^ saveOpenArticleTitleBlock)(NSNotification*) = ^void (NSNotification* notification) {
+        @strongify(self);
+        if (notification.object == self.webView.scrollView) {
+            [self saveOpenArticleTitleWithCurrentlyOnscreenFragment];
+        }
+    };
+
+    [[NSNotificationCenter defaultCenter] addObserverForName:WMF_UIScrollViewDidEndDeceleratingNotification
+                                                      object:nil
+                                                       queue:nil
+                                                  usingBlock:saveOpenArticleTitleBlock];
+    [[NSNotificationCenter defaultCenter] addObserverForName:WMF_UIScrollViewAnimationEndedNotification
+                                                      object:nil
+                                                       queue:nil
+                                                  usingBlock:saveOpenArticleTitleBlock];
+}
+
+- (void)saveOpenArticleTitleWithCurrentlyOnscreenFragment {
+    [[NSUserDefaults standardUserDefaults] wmf_setOpenArticleTitle:[self articleTitleWithCurrentlyOnScreenFragment]];
+}
+
+- (MWKTitle*)articleTitleWithCurrentlyOnScreenFragment {
+    return
+        [[MWKTitle alloc] initWithSite:self.article.title.site
+                       normalizedTitle:self.article.title.text
+                              fragment:[self currentVisibleSection].anchor];
```

Comment:
This will need to be changed a bit for the WKWebView pr, but the same observer approach should carry over.